tests/core/manager/test_middleware_add_and_clear_api.py

```diff
@@ -6,62 +9,158 @@
 )
 
 
-def middleware_factory():
-    def generated_middleware(*args, **kwargs):
-        def middleware(*args, **kwargs):
-            pass
-        return middleware
-    return generated_middleware
-
-
-def test_provider_property_setter_and_getter():
+def test_provider_property_setter_and_getter(middleware_factory):
     provider = BaseProvider()
 
     middleware_a = middleware_factory()
     middleware_b = middleware_factory()
+    middleware_c = middleware_factory()
     assert middleware_a is not middleware_b
+    assert middleware_a is not middleware_c
 
     manager = RequestManager(None, provider, middlewares=[])
 
     assert manager.middlewares == tuple()
 
-    manager.add_middleware(middleware_a)
-    manager.add_middleware(middleware_b)
+    manager.middleware_stack.add(middleware_a)
+    manager.middleware_stack.add(middleware_b)
 
-    manager.clear_middlewares()
+    manager.middleware_stack.clear()
 
     assert manager.middlewares == tuple()
 
-    manager.add_middleware(middleware_b)
-    manager.add_middleware(middleware_a)
-    manager.add_middleware(middleware_b)
-    manager.add_middleware(middleware_b)
-    manager.add_middleware(middleware_a)
+    manager.middleware_stack.add(middleware_c)
+    manager.middleware_stack.add(middleware_b)
+    manager.middleware_stack.add(middleware_a)
+
+    with pytest.raises(ValueError):
+        manager.middleware_stack.add(middleware_b)
 
     assert manager.middlewares == (
         middleware_a,
         middleware_b,
-        middleware_b,
-        middleware_a,
-        middleware_b,
+        middleware_c,
     )
 
 
-def test_modifying_middleware_regenerates_request_functions():
+def test_modifying_middleware_regenerates_request_functions(middleware_factory):
     provider = BaseProvider()
 
     manager = RequestManager(None, provider, middlewares=[])
 
-    id_a = id(manager._wrapped_provider_request_functions)
+    id_start = id(manager._wrapped_provider_request_functions)
+
+    # add
+    manager.middleware_stack.add(middleware_factory(), 'a')
+
+    id_add = id(manager._wrapped_provider_request_functions)
+
+    assert id_add != id_start
+
+    # replace
+    manager.middleware_stack.replace('a', middleware_factory())
+
+    id_replace = id(manager._wrapped_provider_request_functions)
```

Comment:
we don't necessarily need to fix this now, but I'm wondering if using mocks and just asserting that the `_generate_request_functions` method was called after replacing it with a mock.